packages/fxa-auth-server/lib/db/index.ts

```diff
@@ -0,0 +1,1126 @@
+     *
+     * To do a more expensive write that flushes to the underlying
+     * DB, use updateSessionToken instead.
+     */
+    async touchSessionToken(
+      token: any,
+      geo: any,
+      onlyUpdateLastAccessTime = false
+    ) {
+      const { id, uid } = token;
+
+      log.trace('DB.touchSessionToken', { id, uid });
+
+      if (!this.redis || !features.isLastAccessTimeEnabledForUser(uid)) {
+        return;
+      }
+
+      let t;
+      if (onlyUpdateLastAccessTime) {
+        t = {
+          lastAccessTime: token.lastAccessTime,
+          id,
+        };
+      } else {
+        let location;
+        if (geo && geo.location) {
+          location = {
+            city: geo.location.city,
+            country: geo.location.country,
+            countryCode: geo.location.countryCode,
+            state: geo.location.state,
+            stateCode: geo.location.stateCode,
+          };
+        }
+
+        t = {
+          lastAccessTime: token.lastAccessTime,
+          location,
+          uaBrowser: token.uaBrowser,
+          uaBrowserVersion: token.uaBrowserVersion,
+          uaDeviceType: token.uaDeviceType,
+          uaFormFactor: token.uaFormFactor,
+          uaOS: token.uaOS,
+          uaOSVersion: token.uaOSVersion,
+          id,
+        };
+      }
+
+      return this.redis.touchSessionToken(uid, t);
+    }
+
+    /**
+     * Persist updated session-token data to the database.
+     * This is a comparatively expensive call that writes through
+     * to the underlying DB and hence should not be used in
+     * frequently-called routes.
+     *
+     * To do a cheaper write of transient metadata that only hits
+     * redis, use touchSessionToken instead.
+     */
+    async updateSessionToken(sessionToken: any, geo: any) {
+      const { id, uid, lastAccessTime, lastAccessTimeEnabled } = sessionToken;
+
+      // Just for connection pool issue investigation. Make sure the last access time is set to something realistic.
+      log.debug('DB.updateSessionToken', {
+        id,
+        uid,
+        lastAccessTime,
+        lastAccessTimeEnabled,
+      });
+
+      await this.touchSessionToken(sessionToken, geo);
+      await RawSessionToken.update({ id, ...sessionToken });
+    }
+
+    async pruneSessionTokens(uid: string, sessionTokens: any) {
+      log.debug('DB.pruneSessionTokens', {
+        uid,
+        tokenCount: sessionTokens.length,
+      });
+
+      if (
+        !this.redis ||
+        !TOKEN_PRUNING_ENABLED ||
+        !features.isLastAccessTimeEnabledForUser(uid)
+      ) {
+        return;
+      }
+
+      const tokenIds = sessionTokens
+        .filter(
+          (token: any) => token.createdAt <= Date.now() - TOKEN_PRUNING_MAX_AGE
+        )
+        .map((token: any) => token.id);
+
+      if (tokenIds.length === 0) {
+        return;
+      }
+
+      return this.redis.pruneSessionTokens(uid, tokenIds);
+    }
+
+    async createDevice(uid: string, deviceInfo: any): Promise<any> {
```

Comment:
```suggestion
    async createDevice(uid: string, deviceInfo: Pick<
    Device,
    | 'id'
    | 'uid'
    | 'sessionTokenId'
    | 'refreshTokenId'
    | 'name'
    | 'type'
    | 'createdAt'
  > & {
    pushCallback?: string;
    pushPublicKey?: string;
    pushAuthKey?: string;
    availableCommands?: {
      [key: string]: string;
    };
    pushEndpointExpired?:boolean
  }): Promise<Pick<
    Device,
    | 'id'
    | 'uid'
    | 'sessionTokenId'
    | 'refreshTokenId'
    | 'name'
    | 'type'
    | 'createdAt'
  > & {
    pushCallback?: string;
    pushPublicKey?: string;
    pushAuthKey?: string;
    availableCommands?: {
      [key: string]: string;
    };
    pushEndpointExpired?:boolean
  }> {
```

But maybe extract a type definition for reuse.